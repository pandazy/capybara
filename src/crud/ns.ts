import { ns as parentNs } from '../ns';

export function ns(context: string) {
  return parentNs(`crud:${context}`);
}
