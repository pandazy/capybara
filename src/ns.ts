export function ns(context: string) {
  return `Capybara:${context}`;
}
