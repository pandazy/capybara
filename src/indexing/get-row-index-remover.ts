import { keys } from '@pandazy/mole-core/dist/object-methods';
import { CommonMap, Row } from '../types';
import { getIndexKeyGen, hashByDefault } from './get-index-key-gen';
import { makeAssertValidRowFields } from './make-assert-valid-row-fields';
import { ns } from './ns';
import { Index, IndexDef } from './types';

const Context = ns('getRowIndexRemover');

const assertValidRowFields = makeAssertValidRowFields(Context);

export function getRowIndexRemover<TRow extends Row>(
	indexDefs: CommonMap<IndexDef>,
	hash = hashByDefault,
) {
	return (index: Index, row: TRow, rowKey: string) =>
		keys(indexDefs).reduce((updatedIndex: Index, defKey: string) => {
			if (!row) {
				return updatedIndex;
			}

			const { fields } = indexDefs[defKey];
			assertValidRowFields({ fields, row });

			const indexKey = getIndexKeyGen<TRow>(fields, hash)(row);
			const indexVals = updatedIndex[indexKey] || [];
			const recordIdToRemove = indexVals.findIndex(
				(indexedKey) => rowKey === indexedKey,
			);

			if (recordIdToRemove < 0) {
				return updatedIndex;
			}

			const newVals = [...indexVals];
			newVals.splice(recordIdToRemove, 1);
			index[indexKey] = newVals;
			if (newVals.length <= 0) {
				delete updatedIndex[indexKey];
			}
			return updatedIndex;
		}, index);
}
