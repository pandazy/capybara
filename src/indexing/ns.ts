import { ns as parentNs } from '../ns';

export function ns(context: string) {
  return parentNs(`indexing:${context}`);
}
