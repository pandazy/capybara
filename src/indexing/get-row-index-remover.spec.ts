import { getRowIndexRemover } from './get-row-index-remover';
import { ns } from './ns';

interface TestRow {
  name: string;
  job: string;
  alias: string;
}

describe(ns('getRowIndexRemover'), () => {
  it('should remove the index of a given row', () => {
    expect.assertions(3);

    const row: TestRow = {
      name: 'shinobu',
      job: 'Jobless',
      alias: 'kiss-shot',
    };
    const rowKey = '1';
    const fields1 = { fields: ['name', 'job'], isUnique: false };
    const fields2 = { fields: ['alias', 'job'], isUnique: false };
    const customHash = (val: string) => `%${val}%`;
    const index = {
      ['[job:jobless][name:shinobu]']: ['1'],
      ['[alias:kiss-shot][job:jobless]']: ['1', '2'],
    };
    const indexWithHash = {
      '%[job:Jobless][name:shinobu]%': ['1'],
      '%[alias:kiss-shot][job:Jobless]%': ['1', '2'],
    };
    const indexDefs = {
      'job|name': fields1,
      'alias|job': fields2,
    };
    const removeIndex = getRowIndexRemover<TestRow>(indexDefs);
    const removeIndexWithHash = getRowIndexRemover<TestRow>(
      indexDefs,
      customHash,
    );

    expect(removeIndex(index, row, rowKey)).toEqual({
      ['[alias:kiss-shot][job:jobless]']: ['2'],
    });
    expect(removeIndex(index, row, '3')).toEqual(index);
    expect(removeIndexWithHash(indexWithHash, row, rowKey)).toEqual({
      '%[alias:kiss-shot][job:Jobless]%': ['2'],
    });
  });

  it('should throw error if specified field does not exist on row', () => {
    expect.assertions(1);

    const row: TestRow = {
      name: 'shinobu',
      job: 'Jobless',
      alias: 'kiss-shot',
    };

    const fields1 = { fields: ['hobby', 'nationality'], isUnique: false };
    const index = {
      ['[job:jobless][name:shinobu]']: ['1'],
      ['[alias:kiss-shot][job:jobless]']: ['1', '2'],
    };

    const indexDefs = {
      'hobby|nationality': fields1,
    };
    const removeIndex = getRowIndexRemover<TestRow>(indexDefs);

    expect(() =>
      removeIndex(index, row, '3'),
    ).toThrowErrorMatchingSnapshot();
  });
});
