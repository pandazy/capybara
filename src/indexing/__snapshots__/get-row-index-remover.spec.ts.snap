// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capybara:indexing:getRowIndexRemover should throw error if specified field does not exist on row 1`] = `
"
[context]: Capybara:indexing:getRowIndexRemover
[failure cause]: some fields do not exist in the row
[actual value]: {"fields":["hobby","nationality"],"row":{"name":"shinobu","job":"Jobless","alias":"kiss-shot"}}
[actual value type]: Object"
`;
